### PR TITLE
[sw] C logging APIs

### PR DESCRIPTION
**Pull request by sriyerg**
- provide a generic set of APIs to print logs
- provide ability to attach severity (log type) and verbosity to msgs
- provide ability to attach a header to the msg underneath
  - provide ability to construct the header based on severity, verbosity
    file and line number for ease of debug

- provide a generic print_log() function that can parse a formatted string and
take variable number of args

- provide implementation specific APIs in
sw/device/lib/log_uart/log_impl.h to do the prints via UART

- PR #259 is being split to make the review easier
- this PR focuses on providing a high level API and 'typical' SW
implementation of it using UART,
- Subsequent PR will focus on DV specific implementation

This is same as PR #357 with amends. 

Tested the following prints:
```c
  LOG_INFO("info message\n");
  LOG_INFO_LOW("info low\n");
  LOG_INFO_MEDIUM("info med\n");
  LOG_INFO_HIGH("info hi\n");
  LOG_INFO_DEBUG("info debug\n");
  LOG_INFO_FULL("info full\n");
  LOG_WARNING("warn\n");
  LOG_ERROR("err\n");
  LOG_FATAL("fatal\n");

  LOG_INFO("1 in char: %c\n", '1');
  LOG_INFO("1234 in bin: %b\n", 1234);
  LOG_INFO("1234 in dec: %d\n", 1234);
  LOG_INFO("1234 in oct: %o\n", 1234);
  LOG_INFO("1234 in hex: %x\n", 1234);
  LOG_INFO("1234 in negative: %d\n", -1234);
  LOG_INFO("1234 in unsigned: %u\n", -1234);
  LOG_INFO("invalid: %k\n", -1234);
```

Output:
```
Commit ID:  4caa9b7d58e59884414f86f8db6bb863d1104e90
Build Date: 2019-11-07, 18:52:23
INFO: info message
INFO: info low
INFO: info med
INFO: info hi
INFO: info debug
INFO: info full
WARNING: warn
ERROR: err
FATAL: fatal
INFO: 1 in char: 1
INFO: 1234 in bin: 10011010010
INFO: 1234 in dec: 1234
INFO: 1234 in oct: 2322
INFO: 1234 in hex: 4D2
INFO: 1234 in negative: -1234
INFO: 1234 in unsigned: 4294966062
INFO: invalid: [INVALID SPECIFIER: %k]
INFO: Jump!
Hello World! Nov  7 2019 18:43:04
Watch the LEDs!
Try out the switches on the board
or type anything into the console window.
The LEDs show the ASCII code of the last character.
```
Signed-off-by: Srikrishna Iyer <sriyer@google.com>

### CODE REVIEW

**Comment by sriyerg** [comment]
Debugging the meson build failure with @gkelly's help.

File: sw/device/lib/meson.build [hunk 1]
```diff
@@ -118,3 +118,16 @@ sw_lib_usb = declare_dependency(
...
   )
 )
+
+# Message pring library (sw_lib_msg)
+sw_lib_msg = declare_dependency(
+  link_with: static_library(
+    'msg_ot',
+    sources: [
+      'msg_print.c',
+    ]
```
Comment (gkelly) [inline sw/device/lib/meson.build]: Missing comma after `]`.

File: sw/device/lib/msg_print.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+static const char kDecimal = 'd';      // Signed decimal.
+static const char kUnsigned = 'u';     // Unsigned decimal.
+static const char kOctal = 'o';        // Octal.
+static const char kAsciiChar = 'c';    // Single character byte.
+static const char kAsciiStr = 's';     // This is currently not supported.
+static const char kHexUpper = 'X';     // Upper case hex.
+static const char kHexAltUpper = 'H';  // Upper case hex [SystemVerilog style].
+static const char kHexLower = 'x';     // Lower case hex.
+static const char kHexAltLower = 'h';  // Lower case hex [SystemVerilog style].
+static const char kPercent = '%';      // Escape '%'.
```
Comment (mcy) [inline sw/device/lib/msg_print.h]: This is going to duplicate these symbols in every translation unit that includes this file. The correct way to export a constant in C is
```c
// foo.h
extern const char kMyConst;
// foo.c
const char kMyConst = ...;
```

File: sw/device/lib/msg_uart/msg_impl.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+#ifndef _MSG_IMPL_H
+#define _MSG_IMPL_H
+
+#include "sw/device/lib/msg_print.h"
+#include "sw/device/lib/uart.h"
+
+// Stringify stuff.
+#define _xstr(a) _str(a)
+#define _str(a) #a
```
Comment (mcy) [inline sw/device/lib/msg_uart/msg_impl.h]: Give these names that are better scoped to this file, e.g.
```c
#define _MSG_IMPL_STRINGIFY(a) _MSG_IMPL_STRINGIFY(a)
#define _MSG_IMPL_STRINGIFY_INNER(a) #a
```

File: sw/device/lib/msg_uart/msg_impl.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * This macro constructs the message header in the following way:
+ * If MSG_HEADER_INCL_FILE_LINE flag is set:      TYPE (file:line):
+ * If MSG_HEADER_INCL_FILE_LINE flag is not set:  TYPE:
+ *
+ * Message verbosity is currently unused, since the volume of messages
+ * is not expected to be very high.
+ */
+#ifdef MSG_HEADER_INCL_FILE_LINE
+#define MSG_HEADER(msg_type, verbosity) \
+  msg_type verbosity " (" __FILE__ ":" _xstr(__LINE__) "): "
```
Comment (mcy) [inline sw/device/lib/msg_uart/msg_impl.h]: To improve error messages, you can write
```c
"" msg_type "" verbosity " ... "
```
This helps the compiler make sure that msg_type and verbosity are string literals.

File: sw/device/lib/msg_print.c [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    ++fmt;
+    if (ch != kFormatSpecifier) {
+      // Add CR to new line automatically.
+      if (ch == '\n') {
+        print_char('\r');
+      }
+      print_char(ch);
+    } else {
+      int w = 0;
+      while ((ch = *fmt) != '\0') {
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: Factor this loop out into a separate function.

File: sw/device/lib/msg_print.c [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  va_list va;
+
+  va_start(va, fmt);
+  while ((ch = *fmt) != '\0') {
+    ++fmt;
+    if (ch != kFormatSpecifier) {
+      // Add CR to new line automatically.
+      if (ch == '\n') {
+        print_char('\r');
+      }
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: You might want to make sure that if `"\r\n"` is printed, you don't wind up printing `"\r\r\n"`.

File: sw/device/lib/msg_print.c [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      print_char(ch);
+    } else {
+      int w = 0;
+      while ((ch = *fmt) != '\0') {
+        ++fmt;
+        // Parse width field.
+        if (ch >= '0' && ch <= '9') {
+          w = w * 10 + (ch - '0');
+          continue;
+        } else {
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: You don't need this to be in a else block. Prefer early-return style:
```c
if (cond) {
   return/break/continue;
}
// Assume |cond| false here.
```

File: sw/device/lib/msg_print.c [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+          switch (ch) {
+            case '\0': {
+              break;
+            }
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: You probably want to return here. If you break, the outer loop will execute again, but will read past the NUL, which is likely undefined behavior.

File: sw/device/lib/msg_print.c [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  print_char("0123456789ABCDEF"[d]);
+}
+
+static void print_num(print_char_func print_char, int w, unsigned int n,
+                      unsigned int base) {
+  if (--w > 0 || n >= base) {
+    print_num(print_char, w, n / base, base);
+  }
+  print_digit(print_char, n % base);
+}
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: This implementation is both difficult to parse, and wasteful use of the stack. Please re-write it to use a loop, instead. Also, `w` really wants to be named `width`.

File: sw/device/lib/msg_print.c [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+void msg_print(print_char_func print_char, const char *fmt, ...) {
+  char ch;
+  va_list va;
+
+  va_start(va, fmt);
+  while ((ch = *fmt) != '\0') {
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: Assignments in loops... aren't the best. Why not
```c
for (; *fmt != '\0'; ++fmt) {
  char ch = *fmt;
  // ...
}
```
This also helps keep `ch` scoped to where it is used.

File: sw/device/lib/msg_print.c [hunk 11]
```diff
@@ -0,0 +1,107 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/msg_print.h"
+
+#include <stdarg.h>
+
+static void print_digit(print_char_func print_char, unsigned int d) {
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: `d` -> `digit`.

File: sw/device/lib/msg_print.c [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+              unsigned int n = va_arg(va, unsigned int);
+              print_num(print_char, w, n, 16);
+              break;
+            }
+            case kHexUpper:
+            case kHexAltUpper: {
+              unsigned int n = va_arg(va, unsigned int);
+              print_num(print_char, w, n, 16);
+              break;
+            }
```
Comment (mcy) [inline sw/device/lib/msg_print.c]: These are all doing the same thing; re-use the same code block for them, but keep the TODO.

File: sw/device/lib/meson.build (hunk 1 above)
Comment (sriyerg) [reply sw/device/lib/meson.build]: Done.

File: sw/device/lib/msg_print.c (hunk 11 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Done.

File: sw/device/lib/msg_print.c (hunk 9 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Made it inline to reduce stack usage. Will keep the implementation as is and let SW team modify / optimize it as needed (as originally agreed upon), considering this is not going to be used by DV. Added a TODO for this.

File: sw/device/lib/msg_print.c (hunk 10 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Done

File: sw/device/lib/msg_print.c (hunk 6 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Added TODO for this.

File: sw/device/lib/msg_print.c (hunk 5 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Added TODO for this.

File: sw/device/lib/msg_print.h (hunk 2 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.h]: Fixed.

File: sw/device/lib/msg_uart/msg_impl.h (hunk 3 above)
Comment (sriyerg) [reply sw/device/lib/msg_uart/msg_impl.h]: Done.

File: sw/device/lib/msg_uart/msg_impl.h (hunk 4 above)
Comment (sriyerg) [reply sw/device/lib/msg_uart/msg_impl.h]: Done.

File: sw/device/lib/msg_print.c (hunk 7 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Done.

File: sw/device/lib/msg_print.c (hunk 12 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Done.

File: sw/device/lib/msg_print.c (hunk 8 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: Done.

File: sw/device/lib/meson.build (hunk 1 above)
Comment (sriyerg) [reply sw/device/lib/meson.build]: Fixed.

**Comment by sriyerg** [comment]
> Why is this a "messaging" API? This is a logging API, and should probably be named as such, unless you have a very, very, very good compatibility reason for this, in which case it should be documented in a comment.

Done.

**Comment by asb** [comment]
Thank you for doing all of those updates Sri. The only change I'd ask for is some documentation on the supported format strings. Right now, the only way to know that the width field is supported and the set of supported type specifiers (including ones from SystemVerilog) and their meaning is to read `print_log.c`. But it sounds like there's some sort of handover to the software team going on, so I wouldn't be opposed to merging as-is and having the documentation come in a subsequent patch.

**Comment by sriyerg** [comment]
> Thank you for doing all of those updates Sri. The only change I'd ask for is some documentation on the supported format strings. Right now, the only way to know that the width field is supported and the set of supported type specifiers (including ones from SystemVerilog) and their meaning is to read `print_log.c`. But it sounds like there's some sort of handover to the software team going on, so I wouldn't be opposed to merging as-is and having the documentation come in a subsequent patch.

The documentation I added in the sources is quite comprehensive. I assumed with Doxygen style comments, at some point there will be a flow in place to extract it from the sources?

**Comment by sriyerg** [comment]
> k for is some documentation on the supported format strings. Right now, the only way to know that the width field is supported and the set of supported type specifiers (including ones from SystemVerilog) and their meaning is to read `print_log.c`. But it sounds like there's some sort of handover to the software team going on, so I wouldn't be opp



> > Thank you for doing all of those updates Sri. The only change I'd ask for is some documentation on the supported format strings. Right now, the only way to know that the width field is supported and the set of supported type specifiers (including ones from SystemVerilog) and their meaning is to read `print_log.c`. But it sounds like there's some sort of handover to the software team going on, so I wouldn't be opposed to merging as-is and having the documentation come in a subsequent patch.
> 
> The documentation I added in the sources is quite comprehensive. I assumed with Doxygen style comments, at some point there will be a flow in place to extract it from the sources?

Actually, the support if width is very rudimentary and not worth mentioning IMO.

File: sw/device/lib/msg_print.c (hunk 6 above)
Comment (sriyerg) [reply sw/device/lib/msg_print.c]: This is now addressed.

**Comment by sriyerg** [comment]
@gkelly, @mcy, @asb 
PTAL & let me know if there is anything else that needs to be addressed.

**Comment by asb** [comment]
I'll leave this to @gkelly to give a final approving review. It seemed good to me, happy for this to land and for other cleanups etc to come afterwards.

File: sw/device/lib/meson.build [hunk 13]
```diff
@@ -118,3 +118,16 @@ sw_lib_usb = declare_dependency(
     ]
   )
 )
+
+# Message pring library (sw_lib_log)
```
Comment (gkelly) [inline sw/device/lib/meson.build]: `# Logging library that prints to UART directly.`

File: sw/device/lib/log.h [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef _LOG_H_
+#define _LOG_H_
+
+#ifdef DV_SIM
+#include "sw/device/lib/log_uart/log_impl.h"
+#else
+#include "sw/device/lib/log_uart/log_impl.h"
+#endif
```
[the remark this answers is not in the document]
Comment (sriyerg) [reply sw/device/lib/log.h]: Yes, the DV_SIM one will change when that source gets added.

File: sw/device/lib/meson.build (hunk 13 above)
Comment (sriyerg) [reply sw/device/lib/meson.build]: Done.

**Comment by sriyerg** [comment]
Thank you all for your valuable feedback! Merging this in now.

**Pull request merged**